[celeborn] Add celeborn conan dependency

zhangxffff opened the pull request:
<!-- 
Copyright (c) 2025 ByteDance Ltd. and/or its affiliates.

Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

### What problem does this PR solve?
<!--
Please explain the context and the problem.
If this fixes a specific issue, please link it below.
-->
Issue Number: close #36 

### Type of Change
<!-- Please check the one that applies to this PR -->
- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🚀 Performance improvement (optimization)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔨 Refactoring (no logic changes)
- [ ] 🔧 Build/CI or Infrastructure changes
- [ ] 📝 Documentation only

### Description

Pack celeborn c++ client as a conan recipe and add shuffle writer/reader demo with celeborn shuffle client

### Performance Impact
<!--
REQUIRED for Performance PRs and Core Engine changes.
Please verify that your change does not introduce performance regressions.
-->
- [x] **No Impact**: This change does not affect the critical path (e.g., build system, doc, error handling).
- [ ] **Positive Impact**: I have run benchmarks.
    <details>
    <summary>Click to view Benchmark Results</summary>

    ```text
    Paste your google-benchmark or TPC-H results here.
    Before: 10.5s
    After:   8.2s  (+20%)
    ```
    </details>
- [ ] **Negative Impact**: Explained below (e.g., trade-off for correctness).

### Checklist (For Author)
<!--
Please double-check the following before submitting.
-->

- [ ] I have added/updated unit tests (ctest).
- [x] I have verified the code with local build (Release/Debug).
- [x] I have run clang-format / linters.
- [ ] (Optional) I have run Sanitizers (ASAN/TSAN) locally for complex C++ changes.
- [x] No need to test or manual test.

### Breaking Changes
<!--
Does this PR introduce API or ABI incompatibilities?
If yes, please describe how users should migrate.
-->

- [x] No
- [ ] Yes (Description: ...)
    <details>
    <summary>Click to view Breaking Changes</summary>

    ```text
    Breaking Changes:
    - Description of the breaking change.
    - Possible solutions or workarounds.
    - Any other relevant information.
    ```
    </details>

ZacBlanco commented on scripts/conan/recipes/celeborn-cpp-client/all/test_package/test_package.cpp:
```diff
@@ -0,0 +1,38 @@
+      conf::CelebornConf::kClientPushMaxReqsInFlightPerWorker,
+      std::to_string(0));
+
+  std::unique_ptr<PushState> pushState = std::make_unique<PushState>(conf);
```
Is this code enough to properly test the celeborn dependency? Should it try to construct a client/server?

ZacBlanco commented on bolt/shuffle/sparksql/CelebornReaderStreamIterator.cpp:
Could you add some tests for this class?

ZacBlanco commented on bolt/shuffle/sparksql/partition_writer/rss/NativeCelebornClient.cpp:
Could we add some tests for this class?

zhangxffff replied:
`bolt/shuffle/sparksql/tests/CelebornClientTest.cpp` was added to test `NativeCelebornClient` and `CelebornReaderStreamIterator` with mocked celeborn shuffle client.

zhangxffff commented on bolt/shuffle/sparksql/partition_writer/rss/RssClient.h:
```diff
@@ -29,6 +29,7 @@
  * This modified file is released under the same license.
  * --------------------------------------------------------------------------
  */
+#include <cstdint>
```
int32_t and int64_t are used in this header. Adding <cstdint> makes the dependency explicit and avoids relying on indirect includes.

zhangxffff commented on scripts/conan/recipes/fizz/config.yml:
Added a test_package to validate the fizz recipe.

zhangxffff commented on conanfile.py:
```diff
@@ -356,6 +356,7 @@ def requirements(self):
         self.requires("utf8proc/2.11.0", transitive_headers=True, transitive_libs=True)
         self.requires("date/3.0.4-bolt", transitive_headers=True, transitive_libs=True)
         self.requires("libbacktrace/cci.20210118")
+        self.requires("celeborn-cpp-client/main-20251212")
```
Add an option BOLT_ENABLE_CELEBORN_CPP_CLIENT to enable this feature.

ZacBlanco commented on bolt/shuffle/sparksql/CelebornReaderStreamIterator.cpp:
```diff
@@ -0,0 +1,161 @@
+      startMapIndex_(startMapIndex),
+      endMapIndex_(endMapIndex),
+      needCompression_(needCompression) {
+  BOLT_CHECK(shuffleClient_, "ShuffleClient is null for Celeborn reader");
```
nit: consider using `BOLT_CHECK_NOT_NULL`

SteNicholas commented on bolt/shuffle/sparksql/CMakeLists.txt:
```diff
@@ -25,34 +25,58 @@
-    Utils.cpp)
+  PUBLIC arrow::arrow ryu::ryu bolt_vector bolt_core bolt_exec)
+
+if(BOLT_ENABLE_CELEBORN_CPP_CLIENT)
```
Could this branch merge with line 55?

SteNicholas commented on bolt/shuffle/sparksql/partition_writer/rss/NativeCelebornClient.h:
```diff
@@ -0,0 +1,53 @@
+      int32_t numPartitions);
+  ~NativeCelebornClient() override = default;
+
+  int32_t pushPartitionData(int32_t partitionId, char* bytes, int64_t size)
```
```suggestion
  int32_t pushData(int32_t partitionId, char* bytes, int64_t size)
```

zhangxffff replied:
> 

`NativeCelebornClient::pushPartitionData` overrides `RssClient::pushPartitionData`. Since `RssClient` is also used as the base class for the RSS writer in the Gluten Bolt backend, renaming it here would not be straightforward. Maybe we can rename it togather with gluten bolt backend later.

ZacBlanco reviewed (changes requested):
Sorry for the delay. Thank you for addressing the comments. I had just a few more

ZacBlanco commented on bolt/shuffle/sparksql/CMakeLists.txt:
```diff
@@ -25,34 +25,58 @@
+
+add_library(bolt_shuffle_spark_impl ${bolt_shuffle_spark_impl_sources})
+
+target_link_libraries(
```
Thanks for the explanation. My concern is that doing it this way seems to create more options in our build matrix which is already quite complex. It would be easiest from a build and CI perspective to have as few build options as possible, and just run tests with different sets of configuration flags.

I personally would prefer a runtime flags to be able to configure either, but I understand there is already little to no use of the existing shuffle client outside of the tests. I personally don't know what the best approach would be here due to my limited background with bolt's shuffle client.

@markjin1990 do you have any strong opinions about this?

ZacBlanco reviewed (commented):
One last question

zhangxffff commented on bolt/shuffle/sparksql/partition_writer/rss/NativeCelebornClient.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+void NativeCelebornClient::stop() {
+  shuffleClient_->mapperEnd(shuffleId_, mapId_, attemptId_, numMappers_);
```
Thanks for reminding me of this. It’s better to add a defensive check here, and I’ll include it in `stop`.

zhangxffff commented:
@ZacBlanco Many thanks for your review. I’ve removed the `BOLT_ENABLE_CELEBORN_CPP_CLIENT` related option, added a guard to throw an exception if `stop` or `pushData` is called after `NativeCelebornClient::stop`, and added a test case to cover this behavior. Could you please review it again when you have time?

ZacBlanco reviewed (approved):
LGTM
